File: flutter-app/lib/features/events/domain/event.dart
import 'event_status.dart';

/// The Event aggregate as seen by the app (mirrors the event-service `EventDto`).
class Event {
  const Event({
    required this.id,
    required this.status,
    required this.title,
    required this.category,
    required this.eventDate,
    required this.city,
    required this.guestCount,
    required this.ownerId,
    required this.coHostIds,
    required this.memberIds,
    required this.version,
    required this.createdAt,
    required this.updatedAt,
    this.startTime,
    this.endTime,
    this.budgetId,
    this.venueId,
    this.description,
    this.dressCode,
    this.tags,
    this.notes,
  });

  final String id;
  final EventStatus status;
  final String title;
  final String category;

  /// ISO-8601 string as stored by the backend. Use [eventDateTime] for display.
  final String eventDate;
  final String? startTime;
  final String? endTime;
  final String city;
  final int guestCount;
  final String ownerId;
  final List<String> coHostIds;
  final List<String> memberIds;
  final String? budgetId;
  final String? venueId;
  final String? description;
  final String? dressCode;
  final List<String>? tags;
  final String? notes;
  final int version;
  final String createdAt;
  final String updatedAt;

  DateTime? get eventDateTime => DateTime.tryParse(eventDate);

  bool isOwnedBy(String uid) => ownerId == uid;
}

File: backend-api/services/event-service/src/domain/event.ts
import type { DocumentMetadata } from '@hilo/backend-shared';
import type { EventStatus } from './event-status.js';

/**
 * The Event aggregate root (owner: event-service). Combines the mandatory metadata
 * envelope (EOS-000 §35) with event fields (EOS-002-P3-Part-04 §5, EOS-004-P4 §6).
 *
 * The lifecycle state is carried in the metadata `status` field (values from
 * {@link EventStatus}). Ownership/membership are denormalized onto the document
 * (`ownerId`, `coHostIds`, `memberIds`) for O(1) authorization and rule-based reads
 * (EOS-000 §41 — documented denormalization).
 */
export interface Event extends DocumentMetadata {
  status: EventStatus;

  title: string;
  category: string;
  /** ISO-8601 event date. */
  eventDate: string;
  /** ISO-8601 start/end timestamps. */
  startTime?: string;
  endTime?: string;
  city: string;
  guestCount: number;

  ownerId: string;
  coHostIds: string[];
  /** All participants (owner + co-hosts + guests) — powers membership checks/reads. */
  memberIds: string[];

  budgetId?: string;
  venueId?: string;

  description?: string;
  dressCode?: string;
  themeId?: string;
  coverImage?: string;
  tags?: string[];
  notes?: string;
}

/** Whitelisted fields a co-host/owner may patch via update (never ownership/lifecycle). */
export interface EventUpdateInput {
  title?: string;
  category?: string;
  eventDate?: string;
  startTime?: string;
  endTime?: string;
  city?: string;
  guestCount?: number;
  description?: string;
  dressCode?: string;
  tags?: string[];
  notes?: string;
  venueId?: string;
}
